x/cheqd/types/did_validation.go

```diff
@@ -0,0 +1,220 @@
+// 5 - #([^#]+[\$]?) - group for fragment, starts with #, includes #                              (#key1???)
+// 6 - [^#]+[\$]?    - fragment only															  (key1???)
+// {0,1} - means that number of fragments can be only 0 or 1
+// Amount of query is not limited.
+var SplitDIDURL, _           = regexp.Compile(`([^/?#]*)?([^?#]*)(\?([^#]*)){0,1}(#([^#]+$)){0,1}$`)
+var DIDPathAbemptyRegexp, _  = regexp.Compile(`^[/a-zA-Z0-9\-\.\_\~\%\!\$\&\'\(\)\*\+\,\;\=\:\@]+$`)
+var DIDQueryRegexp, _        = regexp.Compile(`^[/a-zA-Z0-9\-\.\_\~\%\!\$\&\'\(\)\*\+\,\;\=\:\@\/\?]*$`)
+var DIDFragmentRegexp, _     = regexp.Compile(`^[/a-zA-Z0-9\-\.\_\~\%\!\$\&\'\(\)\*\+\,\;\=\:\@\/\?]*$`)
+
+//[^/]+(/[^\?#]*)?(\?[^#]+)?(#.+)?
+//Old implementation
+//var DidForbiddenSymbolsRegexp, _ = regexp.Compile(`^[^#?&/\\]+$`)
+//
+//func SplitDidUrlIntoDidAndFragment(didUrl string) (string, string) {
+//	fragments := strings.Split(didUrl, "#")
+//	return fragments[0], fragments[1]
+//}
+//
+//func IsDidFragment(prefix string, didUrl string) bool {
+//	if !strings.Contains(didUrl, "#") {
+//		return false
+//	}
+//
+//	if didUrl[0] == '#' {
+//		return true
+//	}
+//
+//	did, _ := SplitDidUrlIntoDidAndFragment(didUrl)
+//	return IsValidDid(prefix, did)
+//}
+//
+//func IsFullDidFragment(prefix string, didUrl string) bool {
+//	if !strings.Contains(didUrl, "#") {
+//		return false
+//	}
+//
+//	did, _ := SplitDidUrlIntoDidAndFragment(didUrl)
+//	return IsValidDid(prefix, did)
+//}
+//
+//func ResolveId(did string, methodId string) string {
+//	result := methodId
+//
+//	methodDid, methodFragment := SplitDidUrlIntoDidAndFragment(methodId)
+//	if len(methodDid) == 0 {
+//		result = did + "#" + methodFragment
+//	}
+//
+//	return result
+//}
+//
+//func IsNotValidDIDArray(prefix string, array []string) (bool, int) {
+//	for i, did := range array {
+//		if !IsValidDid(prefix, did) {
+//			return true, i
+//		}
+//	}
+//
+//	return false, 0
+//}
+//
+//func IsNotValidDIDArrayFragment(prefix string, array []string) (bool, int) {
+//	for i, did := range array {
+//		if !IsDidFragment(prefix, did) {
+//			return true, i
+//		}
+//	}
+//
+//	return false, 0
+//}
+//
+//func IsValidDid(prefix string, did string) bool {
+//	if len(did) == 0 {
+//		return false
+//	}
+//
+//	if !DidForbiddenSymbolsRegexp.MatchString(did) {
+//		return false
+//	}
+//
+//	// FIXME: Empty namespace must be allowed even if namespace is set in state
+//	// https://github.com/cheqd/cheqd-node/blob/main/architecture/adr-list/adr-002-cheqd-did-method.md#method-specific-identifier
+//	return strings.HasPrefix(did, prefix)
+//}
+
+
+// DID
+
+func ValidateDID(did string, method string, allowedNamespaces []string) error {
+	method, namespace, unique_id := SplitDID(did)
+
+	// check method
+	if method != method {
+		return ErrStaticDIDBadMethod.Wrap(method)
+	}
+	// check namespaces
+	if !DidNamespaceRegexp.MatchString(namespace) || !utils.Contains(allowedNamespaces, namespace) {
```

Comment:
Let's move regexp matches to the split method

Reply:
So validate will only compare namespace / method with passed parameters

Reply:
Will it be cool to throw the exception while splitting the DID? From my point of view it's a phase of validation.